Source/cmake/OptionsCommon.cmake

```diff
@@ -154,7 +154,7 @@ if (USE_THIN_ARCHIVES)
 endif ()
 
 set(ENABLE_DEBUG_FISSION_DEFAULT OFF)
-if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+if (CMAKE_BUILD_TYPE STREQUAL "Debug" OR CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo" OR "${CMAKE_CXX_FLAGS}" MATCHES "-g" OR "${CMAKE_CXX_FLAGS_RELEASE}" MATCHES "-g")
```

Comment:
In particular, this doesn't work if build type is MinSizeRel or something custom.

Thinking about this more, we should definitely test for "RelWithDebInfo" build type, but I wonder if it's really valuable to add -gsplit-dwarf in response to a manually-added -g, since anyone doing that manually could also manually add -gsplit-dwarf. No strong opinion... as long as it doesn't break distro debuginfo, it's fine.

Reply:
Can you r+ then?

Reply:
Almost. If you test  `${CMAKE_CXX_FLAGS_RELEASE}`, then you should at least also test ``${CMAKE_CXX_FLAGS_MINSIZEREL}` as well, since that is the other standard build type that doesn't normally have debuginfo. I would r+ that.

But there is a disadvantage: you might be testing the flags for a build type that's not actually in use. E.g. `${CMAKE_CXX_FLAGS_RELEASE}` might really contain -g, but the user didn't select the Release build type. So best would be to try testing `${CMAKE_CXX_FLAGS_$<CONFIG>}` instead of `${CMAKE_CXX_FLAGS_RELEASE}`. I'm not certain if that will work, but if it does, then you have a simpler test that will check the flags for the configuration that's actually in use and none other and will work even for custom build types. But if that doesn't work, then don't worry about it.

Reply:
There is no need to use the `$<CONFIG>` generator expression, there is `${CMAKE_BUILD_TYPE}`, and CMake supports indirect expansion of variables, so things like this work:

```cmake
string(TOUPPER "${CMAKE_BUILD_TYPE}" BUILD_TYPE_IN_CAPS)
message("Current build type CXX flags: ${CMAKE_CXX_FLAGS_${BUILD_TYPE_IN_CAPS}}")
```

Then, you want to use [separate_arguments()](https://cmake.org/cmake/help/v3.12/command/separate_arguments.html) instead of plain string matching, then check whether any of the items in the resulting list matches `-g<number>`, `-gdwarf`, `-gdwarf-<version>`, or `-ggdb`. With a plain regexp matching on `-g`, you will be matching other options that tweak how debuginfo is produced but have nothing to do with enabling debuginfo, and also other non-debuginfo compiler options like `-gen-decls` or `-gtoggle`...

So let's not go down this rabbit hole, or we would need to reimplement the same logic for command line parsing that the compiler does.

Instead, let's remove the check on the build type, and *always* add `-gsplit-dwarf` to the flags when supported by the compiler and linker. Why? Because this _modifies_ how debuginfo is generated, but very importantly does _not_ enable debuginfo generation, so it will be a no-op if there is no `-g...` option that enables debuginfo generation.

The [GCC manual](https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/Debugging-Options.html#Debugging-Options) is quite clear about this (emphasis mine):

> `-gsplit-dwarf`
> **If DWARF debugging information is enabled**, separate as much debugging information as possible into a separate output file with the extension .dwo. This option allows the build system to avoid linking files with debug information. To be useful, this option requires a debugger capable of reading .dwo files.

TL;DR: Always pass `-gsplit-dwarf` if supported, regardless of the build type, the compiler already handles this for us.

Reply:
Ah... good idea. :P

(Assuming this doesn't break distros, which we'll find out soon enough.)

Reply:
If that happened, we can change things so instead of _appending_ the flags, we would _prepend_ them, and distros can still override our default by setting `C[XX]FLAGS=-gno-split-dwarf` in the environment before invoking CMake. That will result in both flags in the command line, as in `-gsplit-dwarf -foo -bar -gno-split-dwarf` but the flag provided by the distro packager would take precedence.

Reply:
Um, good point: the rest of the code in this file is very careful to prepend rather than append almost always, and only ever append in cases where WebKit absolutely needs a certain flag (like `-fno-exceptions` or `-fno-rtti`). So the implementation here is backwards and should be flipped. Preexisting problem, though.